worker/upload.go

```diff
@@ -0,0 +1,836 @@
+package worker
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"sort"
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	rhpv2 "go.sia.tech/core/rhp/v2"
+	"go.sia.tech/core/types"
+	"go.sia.tech/renterd/api"
+	"go.sia.tech/renterd/object"
+	"go.sia.tech/renterd/tracing"
+	"lukechampine.com/frand"
+)
+
+// TODO: add priority queues
+
+const defaultAvgUpdateInterval = 10 * time.Second
+
+var errNoQueue = errors.New("no queue")
+
+type (
+	uploadID [8]byte
+
+	uploader struct {
+		w *worker
+
+		statsOverdrive *average
+		statsStopChan  chan struct{}
+
+		poolMu sync.Mutex
+		pool   []*uploadQueue
```

Comment:
Why a slice instead of a map from fcid to queue?

Also maybe we can have a slightly more descriptive name than `pool`. Wasn't sure what kind of thing we are pooling here but I guess 1 queue per host?

Reply:
Because I sort it. Also because I in-line replace them when I update the pool but I guess I can easily make it a map. I don't see why it can't be a slice though, making it a map complicates things when I fetch a queue I think.